spring-batch-core/src/main/java/org/springframework/batch/core/repository/JobRepository.java

```diff
@@ -178,14 +178,28 @@ JobExecution createJobExecution(String jobName, JobParameters jobParameters)
 	 */
 	void updateExecutionContext(JobExecution jobExecution);
 
+	/**
+	 * @return all step executions, from all job executions, for the given job instance.
+	 */
+	Collection<StepExecution> getStepExecutions(JobInstance jobInstance);
```

Comment:
We can use Java 8 "default method in interface" for backward compatibility here. For example:

```java
default Collection<StepExecution> getStepExecutions(JobInstance jobInstance) {
   throw new UnsupportedOperationException();
}
```